### PR TITLE
Fix ghost BlockStorage blocks due to iron golems, snow golems and wither spawning

### LINKED ISSUE #2425: Iron Golem creation using Iron Golem Assembler
## :round_pushpin: Description (REQUIRED)

The Iron Golem creation works from Iron Golem Assembler, and this cause ghost blocks.

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)

1. Build Iron Golem from Iron Golem Assembler and Curved Pumpkin.

https://www.youtube.com/watch?v=ay4D5EWaqO0

## :bulb: Expected behavior (REQUIRED)

The Iron Golem creation not work from Iron Golem Assembler.

## :open_file_folder: /error-reports/ Folder

No error report.

## :compass: Environment (REQUIRED)

[16:37:52 INFO]: This Server uses the following setup of Slimefun:
[16:37:52 INFO]: Paper v1_16_R2
[16:37:52 INFO]: CS-CoreLib vDEV - 91 (git e42a05d1)
[16:37:52 INFO]: Slimefun vDEV - 708 (git de5db265)
[16:37:52 INFO]: Metrics build: #17
[16:37:52 INFO]:
[16:37:52 INFO]: Installed Addons: (0)

### PR DESCRIPTION
**Pull request by md5sha256**
## Description
Fix ghost `BlockStorage` blocks due to players building/spawning iron golems, snow golems, and withers.\

Fixes 2425 (tagged below)

## Proposed changes

Adds a new listener (`CreatureBuildListener`) which cancels `CreatureSpawnEvents` if one of the blocks used to build 
either an iron golem, snow golem (EntityType#SNOWMAN), or wither is tracked by `BlockStorage` (`BlockStorage#hasInfo == true`)

## Related Issues (if applicable)

Resolves #2425

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.19.*).
- [ ] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
 import javax.annotation.Nullable;
 import javax.annotation.ParametersAreNonnullByDefault;
 
+import io.github.thebusybiscuit.slimefun4.implementation.listeners.entity.CreatureBuildListener;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/Slimefun.java:18]: Going to want to move this import to be with the other listener imports

**Comment by md5sha256** [comment]
Just a note, I haven't completed a full test with Slimefun/addons however I've manually tested the core logic of parsing the removed blocks (So `implementation/utils/blockpattern/TShapedBlockPattern` and `implementation/utils/blockpattern/WitherBuildPattern`) on a dev-plugin in my local workspace. I'll submit proper tests (and add unit tests) when I have a bit more time, however, I would like some sort of feedback.

**Comment by TheBusyBiscuit** [comment]
That's awesome!
Good and clean solution to a very awkward problem, thank you!

Does this also prevent the blocks from being removed or will it just cancel the mob spawning?
If not, then we may need to restore the blocks manually.

**Comment by md5sha256** [comment]
> That's awesome! Good and clean solution to a very awkward problem, thank you!
> 
> Does this also prevent the blocks from being removed or will it just cancel the mob spawning? If not, then we may need to restore the blocks manually.

This just cancels the mob spawning as I'm not sure how to deal with whether or not the `BlockStorage` entry should be removed. The logic can definitely be changed to remove the `BlockStorage` entry however.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/entity/CreatureBuildListener.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            Block base = location.getBlock();
+            removedBlocks = Arrays.asList(base, base.getRelative(BlockFace.UP), base.getRelative(BlockFace.UP, 2));
+        } else if (event.getEntityType() == EntityType.WITHER) {
+            // Add the soul sand and wither skulls
+            removedBlocks = WitherBuildPattern.getMatchingBlocks(location);
+        } else {
+            // This should not happen as we checked the SpawnReason earlier
+            // This return statement is just to make the compiler happy
+            return;
+        }
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/entity/CreatureBuildListener.java:56]: Might be worth it to use a switch statement here

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/entity/CreatureBuildListener.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onCreatureSpawn(CreatureSpawnEvent event) {
+        // We only care about building iron golems, snowmen and withers
+        if (event.getSpawnReason() != CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM
+                && event.getSpawnReason() != CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN
+                && event.getSpawnReason() != CreatureSpawnEvent.SpawnReason.BUILD_WITHER) {
+            return;
+        }
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/entity/CreatureBuildListener.java:37]: I'd recommend adding some empty lines, looks nicer and more readability

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return Arrays.asList(center, north, south);
+    }
+
+    /**
+     * Gets the blocks which are in a 3-block line facing east-west
+     *
+     * @param center The block at the center of the line
+     * @return Returns a {@link Collection} of {@link Block}s comprised of the 3 blocks
+     */
+    public static Collection<Block> getLineEastWest(Block center) {
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:114]: Nonnull

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+/**
+ * This utility class holds methods relating to a pattern of blocks in a T-shape
+ *
+ * @author md5sha256
+ */
+@ParametersAreNonnullByDefault
+public class TShapedBlockPattern {
+
+    private TShapedBlockPattern() {
+    }
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:24]: This should be one line

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ * &nbsp #<br>
+ * <br>
+ *
+ * @author md5sha256
+ */
+@ParametersAreNonnullByDefault
+public class WitherBuildPattern {
+
+    private WitherBuildPattern() {
+    }
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java:31]: This should be one line

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+     * @param location The base of the wither structure
+     * @return Returns the line of blocks facing north-south
+     */
+    private static @Nonnull Collection<Block> getWitherHeadsNorthSouth(Location location) {
+        Block base = location.getBlock();
+        Block center = base.getRelative(BlockFace.UP, 2);
+        return TShapedBlockPattern.getLineNorthSouth(center);
+    }
+
+
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java:93]: 2 empty lines

**Comment by TheBusyBiscuit** [comment]
I meaan the physical blocks, not the block storage data.
Like, will this leave the structure behind or will it create ghost blocks and air nonetheless.

**Review by JustAHuman-xD — CHANGES_REQUESTED** [review]

**Comment by md5sha256** [comment]
> I meaan the physical blocks, not the block storage data. Like, will this leave the structure behind or will it create ghost blocks and air nonetheless.

This will leave the structure behind, no physical blocks will be removed.

**Comment by TheBusyBiscuit** [comment]
> > I meaan the physical blocks, not the block storage data. Like, will this leave the structure behind or will it create ghost blocks and air nonetheless.
> 
> This will leave the structure behind, no physical blocks will be removed. EDIT: The blocks are only removed by vanilla/craftbukkit if the `CreatureSpawnEvent` isn't canceled (I looked into the internals + verified by running a test in-game). Therefore, since the listener will cancel the event, the blocks won't be removed.

Perfect! That's what I wanted to know.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/entity/CreatureBuildListener.java [hunk 8]
```diff
@@ -35,24 +34,29 @@ public void onCreatureSpawn(CreatureSpawnEvent event) {
...
+                // Add the two snow blocks and the carved pumpkin head
+                Block base = location.getBlock();
+                removedBlocks = Arrays.asList(base, base.getRelative(BlockFace.UP), base.getRelative(BlockFace.UP, 2));
+            }
+            case WITHER ->
+                // Add the soul sand and wither skulls
+                removedBlocks = WitherBuildPattern.getMatchingBlocks(location);
+            default -> {
+                // This should not happen as we checked the SpawnReason earlier
+                // This return statement is just to make the compiler happy
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/entity/CreatureBuildListener.java:57]: Might be worth adding a WARNING log here, as in "unexpected entity type X spawned through an unknown build pattern." or something.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Also don't forget to mark params are nonnull and add Validate checks

**Comment by md5sha256** [comment]
> Also don't forget to mark params are nonnull and add Validate checks

I added the class-level annotation `ParamsAreNonnullByDefault`, is that accepted or must I annotate the params individually?

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+     *
+     * @param material The {@link Material} which the blocks should match
+     * @param location The {@link Location} of the lowest block at the center, aka the base block of the T-shape
+     * @return Returns non-null {@link Collection} of {@link Block}s which match the specified {@link Material} and are
+     * in a T-Shape in no particular order
+     */
+    public static @Nonnull Collection<Block> getMatchingBlocks(Material material, Location location) {
+        Validate.notNull(material, "Material cannot be null");
+        Validate.notNull(location, "Location cannot be null");
+        Validate.notNull(location.getWorld(), "Location#getWorld cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:43]: Use Preconditions as Validate only affects tests.

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    /**
+     * Check whether all the {@link Block}s in a {@link Collection} are of a specified {@link Material}
+     *
+     * @param material The material to check
+     * @param blocks   THe blocks to check
+     * @return True if all blocks are of the specified material, false otherwise
+     */
+    public static boolean allBlocksMatchMaterial(Material material, Collection<Block> blocks) {
+        Validate.notNull(material, "Material cannot be null");
+        Validate.notNull(blocks, "blocks cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:64]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    /**
+     * Get the blocks which make up at T-shape facing east-west
+     *
+     * @param location The {@link Location} of the lowest block at the center, aka the base block of the T-shape
+     * @return Returns a {@link Collection} of {@link Block}s which match the T-shape
+     */
+    public static @Nonnull Collection<Block> getTShapeEastWest(Location location) {
+        Validate.notNull(location, "Location cannot be null");
+        Validate.notNull(location.getWorld(), "Location#getWorld cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:81]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+    /**
+     * Get the blocks which make up at T-shape facing north-south
+     *
+     * @param location The {@link Location} of the lowest block at the center, aka the base block of the T-shape
+     * @return Returns a {@link Collection} of {@link Block}s which match the T-shape
+     */
+    public static @Nonnull Collection<Block> getTShapeNorthSouth(Location location) {
+        Validate.notNull(location, "Location cannot be null");
+        Validate.notNull(location.getWorld(), "Location#getWorld cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:97]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    /**
+     * Gets the blocks which are in a 3-block line facing north-south
+     *
+     * @param center The block at the center of the line
+     * @return Returns a {@link Collection} of {@link Block}s comprised of the 3 blocks
+     */
+    public static @Nonnull Collection<Block> getLineNorthSouth(Block center) {
+        Validate.notNull(center, "center cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:112]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    /**
+     * Gets the blocks which are in a 3-block line facing east-west
+     *
+     * @param center The block at the center of the line
+     * @return Returns a {@link Collection} of {@link Block}s comprised of the 3 blocks
+     */
+    public static @Nonnull Collection<Block> getLineEastWest(Block center) {
+        Validate.notNull(center, "center cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:125]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java:13]:
```suggestion
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import org.apache.commons.lang.Validate;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+     * <br>
+     * <strong>This method should not be used to test if a Wither can be spawned at this location as it does not
+     * check whether the other blocks within the pattern are air blocks.</strong>
+     *
+     * @param location The {@link Location} of the lowest block at the center, aka the base block of the build pattern
+     * @return Returns non-null {@link Collection} of {@link Block}s which are eligible for creating a wither
+     */
+    public static @Nonnull Collection<Block> getMatchingBlocks(Location location) {
+        Validate.notNull(location, "Location cannot be null");
+        Validate.notNull(location.getWorld(), "Location#getWorld cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java:48]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * Get the blocks which should contain wither heads if the structure is facing east-west
+     *
+     * @param location The base of the wither structure
+     * @return Returns the line of blocks facing east-west
+     */
+    private static @Nonnull Collection<Block> getWitherHeadsEastWest(Location location) {
+        Validate.notNull(location, "Location cannot be null");
+        Validate.notNull(location.getWorld(), "Location#getWorld cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java:79]: Preconditions

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * Get the blocks which should contain wither heads if the structure is facing north-south
+     *
+     * @param location The base of the wither structure
+     * @return Returns the line of blocks facing north-south
+     */
+    private static @Nonnull Collection<Block> getWitherHeadsNorthSouth(Location location) {
+        Validate.notNull(location, "Location cannot be null");
+        Validate.notNull(location.getWorld(), "Location#getWorld cannot be null");
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/WitherBuildPattern.java:93]: Preconditions

File: src/test/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TestCreatureBuildListener.java [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
```
Comment (JustAHuman-xD) [inline src/test/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/TestCreatureBuildListener.java:36]:
```suggestion
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.EnumSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Stream;

import javax.annotation.Nonnull;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.entity.EntityType;
import org.bukkit.event.entity.CreatureSpawnEvent;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;
import io.github.thebusybiscuit.slimefun4.implementation.listeners.entity.CreatureBuildListener;
import io.github.thebusybiscuit.slimefun4.utils.blockpattern.TShapedBlockPattern;

import me.mrCookieSlime.Slimefun.api.BlockStorage;

import be.seeseemelk.mockbukkit.MockBukkit;
import be.seeseemelk.mockbukkit.ServerMock;
import be.seeseemelk.mockbukkit.WorldMock;
import be.seeseemelk.mockbukkit.entity.LivingEntityMock;
import be.seeseemelk.mockbukkit.entity.SimpleMobMock;
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TestTShapedBlockPattern.java [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
```
Comment (JustAHuman-xD) [inline src/test/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TestTShapedBlockPattern.java:22]:
```suggestion
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;

import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;

import be.seeseemelk.mockbukkit.MockBukkit;
import be.seeseemelk.mockbukkit.WorldMock;
import be.seeseemelk.mockbukkit.block.BlockMock;
```

File: src/test/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TestWitherBuildPattern.java [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
```
Comment (JustAHuman-xD) [inline src/test/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TestWitherBuildPattern.java:16]:
```suggestion
import java.util.ArrayList;
import java.util.Collection;

import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;

import be.seeseemelk.mockbukkit.MockBukkit;
import be.seeseemelk.mockbukkit.WorldMock;
```

**Review by JustAHuman-xD — CHANGES_REQUESTED** [review]

**Comment by J3fftw1** [comment]
Marking this as stale.
Would be a nice fix to have in :D

**Comment by md5sha256** [comment]
Totally forgot about this, will push some changes in the next few days

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java (hunk 9 above)
Comment (md5sha256) [reply src/main/java/io/github/thebusybiscuit/slimefun4/utils/blockpattern/TShapedBlockPattern.java:45]: Does validate only affect tests? Looking at the src it just throws regardless? I'll change it anyway but im a bit confused

**Comment by md5sha256** [comment]
I'm gonna leave a comment here before I push the commit which changes Validate for Preconditions. I can see Validate being used in other SF classes (ex: `package io.github.thebusybiscuit.slimefun4.utils.biomes.BiomeMap`), just making sure we do indeed want to use Preconditions. @JustAHuman-xD

**Comment by JustAHuman-xD** [comment]
> I'm gonna leave a comment here before I push the commit which changes Validate for Preconditions. I can see Validate being used in other SF classes (ex: `package io.github.thebusybiscuit.slimefun4.utils.biomes.BiomeMap`), just making sure we do indeed want to use Preconditions. @JustAHuman-xD 

We do, validate only runs on tests as we have found, there's an open issue about it iirc, new code uses preconditions until a better thing is created in dough

**Comment by md5sha256** [comment]
can we also unmark this as stale?

**Comment by variananora** [comment]
Just tested this on 1.20.1, there a few things that I found:

1. Snow golem will not spawn in any of the possible spawning rules (normal, upside down, or side)
2. Spawning iron golem in the default normal T position will not spawn the mob
3. Spawning iron golem side ways or upside down did spawn the mob and leaving ghost blocks behind.
4. Spawning wither using soul sand will not spawn the mob
5. Spawning wither using soul soil will spawn the mob and leaving ghost blocks behind.

Video: https://youtu.be/qaD1JPAZJqI

P.S. [Here](https://cdn.discordapp.com/attachments/1135126914529513513/1135126940932644944/SlimefunTestAddon-3796.jar) is a little addon to add snow block, soul sand, and soul soil for spawning those mob. (Use it on your own risk)

P.S.S. If you need me to test the build, feel free to ping me on Discord or Github and flag the PR "Needs Testing", I will look into it later!

**Comment by J3fftw1** [comment]
the edge cases you have found seem to be edge cases where spigot doesnt even check for.

**Comment by md5sha256** [comment]
> the edge cases you have found seem to be edge cases where spigot doesnt even check for.

does paper check for these?

**Comment by md5sha256** [comment]
> Just tested this on 1.20.1, there a few things that I found:
> 
> 1. Snow golem will not spawn in any of the possible spawning rules (normal, upside down, or side)
> 2. Spawning iron golem in the default normal T position will not spawn the mob
> 3. Spawning iron golem side ways or upside down did spawn the mob and leaving ghost blocks behind.
> 4. Spawning wither using soul sand will not spawn the mob
> 5. Spawning wither using soul soil will spawn the mob and leaving ghost blocks behind.
> 
> Video: https://youtu.be/qaD1JPAZJqI
> 
> P.S. [Here](https://cdn.discordapp.com/attachments/1135126914529513513/1135126940932644944/SlimefunTestAddon-3796.jar) is a little addon to add snow block, soul sand, and soul soil for spawning those mob. (Use it on your own risk)
> 
> P.S.S. If you need me to test the build, feel free to ping me on Discord or Github and flag the PR "Needs Testing", I will look into it later!

interesting, i'm very surprised it is not working as expected, let me have another crack

**Comment by J3fftw1** [comment]
> > the edge cases you have found seem to be edge cases where spigot doesnt even check for.
> 
> does paper check for these?

dont know dont think they do var most likely tested on paper

**Comment by variananora** [comment]
> dont know dont think they do var most likely tested on paper

Yeah I exclusively tested everything on paper these days

**Comment by md5sha256** [comment]
Ready for another round of testing. Fixes the issues with inverted orientations and the use of soul sand. Also added more unit tests for the new cases.
cc: @variananora for a re-test :)
